pkg/agent/manager/comm.go

```diff
@@ -0,0 +1,236 @@
+	}
+	m.mtx.Unlock()
+
+	client, err := m.newClient(svid, key)
+	if err != nil {
+		return err
+	}
+
+	for _, id := range spiffeIDs {
+		m.syncClients.add(id, client)
+	}
+
+	return nil
+}
+
+func (p *clientsPool) add(spiffeID string, client *client) {
+	// If there is already a connection with the specified spiffeID, close it first.
+	if c := p.get(spiffeID); c != nil {
+		c.close()
+	}
+
+	p.m.Lock()
+	defer p.m.Unlock()
+	p.clients[spiffeID] = client
+}
+
+func (p *clientsPool) get(spiffeID string) *client {
+	p.m.Lock()
+	defer p.m.Unlock()
+	return p.clients[spiffeID]
+}
+
+func (m *manager) getRotationClient() *client {
+	return m.syncClients.get(rotatorTag)
+}
+
+func (m *manager) renewRotatorClient() error {
+	svid, key := m.getBaseSVIDEntry()
+	c, err := m.newClient(svid, key)
+	if err != nil {
+		return err
+	}
+	m.syncClients.add(rotatorTag, c)
+	return nil
+}
+
+// close releases the pool's resources.
+func (p *clientsPool) close() {
+	p.m.Lock()
+	defer p.m.Unlock()
+	for _, c := range p.clients {
+		c.close()
+	}
+}
+
+func (c *client) close() {
+	c.conn.Close()
+}
+
+type update struct {
+	regEntries map[string]*common.RegistrationEntry
+	svids      map[string]*node.Svid
+	lastBundle []byte
+}
+
+func (u *update) String() string {
+	var buffer bytes.Buffer
+	buffer.WriteString("{ regEntries: [")
+	for _, re := range u.regEntries {
+		buffer.WriteString("{ spiffeID: ")
+		buffer.WriteString(re.SpiffeId)
+		buffer.WriteString(", parentID: ")
+		buffer.WriteString(re.ParentId)
+		buffer.WriteString(", selectors: ")
+		buffer.WriteString(fmt.Sprintf("%v", re.Selectors))
+		buffer.WriteString("}")
+	}
+	buffer.WriteString("], svids: [")
+	for key, svid := range u.svids {
+		buffer.WriteString(key)
+		buffer.WriteString(": ")
+		buffer.WriteString(svid.String()[:30])
+		buffer.WriteString(" ")
+	}
+	buffer.WriteString("], lastBundle: ")
+	if u.lastBundle != nil && len(u.lastBundle) > 0 {
+		buffer.WriteString("bytes")
+	} else {
+		buffer.WriteString("none")
+	}
+	buffer.WriteString("}")
+	return buffer.String()
+}
+
+func (c *client) sendAndReceive(r *node.FetchSVIDRequest) (*update, error) {
+	// Enable pipelined access by client to this function.
+	go func() {
+		c.requests <- r
+	}()
+	req := <-c.requests
+
+	var stream node.Node_FetchSVIDClient
+	var err error
+	// Create a new stream, we retry some times because under certain conditions
+	// the stream cannot be created because it throws an "all SubConns are in TransientFailure"
+	// which can be recovered retrying. Some times it is not possible to avoid that error,
+	// which could be because the server is down, or because there are some problem with the
+	// connection's mTLS configuration.
+	for retries := 0; retries < 5; retries++ {
+		stream, err = c.nodeClient.FetchSVID(context.Background())
+		if err == nil {
+			break
+		}
+		c.log.Errorf("failed to get stream: %v. try number: %d", err, retries)
```

Comment:
do we need to sleep or anything between these tries?

Reply:
Maybe...not sure, but I think it would be a good idea to sleep between those lines. How many seconds do you think would be good? (retries + 1) seconds makes sense? This way we sleep more on each iteration, let me know what you think.